File: base/src/main/vsdk/toolkit/processing/linealAlgebra/LinearAlgebraEngine.java
package vsdk.toolkit.processing.linealAlgebra;

import vsdk.toolkit.common.linealAlgebra.MatrixNxM;

public class LinearAlgebraEngine
{
    private static final LinearAlgebraEngine DEFAULT_ENGINE =
        fromStrategy(StrategySelector.ComputeStrategy.NAIVE_COFACTOR_CPU);

    private final LinearAlgebraBackend backend;

    public LinearAlgebraEngine(LinearAlgebraBackend backend)
    {
        if ( backend == null ) {
            throw new IllegalArgumentException("backend cannot be null");
        }
        this.backend = backend;
    }

    public static LinearAlgebraEngine defaultEngine()
    {
        return DEFAULT_ENGINE;
    }

    public static LinearAlgebraEngine fromStrategy(StrategySelector.ComputeStrategy strategy)
    {
        StrategySelector selector = new StrategySelector();
        return new LinearAlgebraEngine(selector.select(strategy));
    }

    public double determinant(MatrixNxM matrix)
    {
        return backend.determinantStrategy().determinant(matrix);
    }

    public MatrixNxM inverse(MatrixNxM matrix)
    {
        return backend.inverseStrategy().inverse(matrix);
    }

    public LinearAlgebraBackend backend()
    {
        return backend;
    }

    static final class DefaultBackend implements LinearAlgebraBackend
    {
        private final DeterminantStrategy determinant;
        private final InverseStrategy inverse;

        DefaultBackend(Object strategy)
        {
            if ( !(strategy instanceof DeterminantStrategy) ) {
                throw new IllegalArgumentException("strategy must implement DeterminantStrategy");
            }
            if ( !(strategy instanceof InverseStrategy) ) {
                throw new IllegalArgumentException("strategy must implement InverseStrategy");
            }
            this.determinant = (DeterminantStrategy) strategy;
            this.inverse = (InverseStrategy) strategy;
        }

        @Override
        public String id()
        {
            return determinant.id();
        }

        @Override
        public DeterminantStrategy determinantStrategy()
        {
            return determinant;
        }

        @Override
        public InverseStrategy inverseStrategy()
        {
            return inverse;
        }
    }
}

File: typescript/base/src/main/vsdk/toolkit/processing/linealAlgebra/LinearAlgebraEngine.ts
import type { MatrixNxM } from "../../common/linealAlgebra/MatrixNxM.js";
import type { DeterminantStrategy } from "./DeterminantStrategy.js";
import type { InverseStrategy } from "./InverseStrategy.js";
import type { LinearAlgebraBackend } from "./LinearAlgebraBackend.js";
import { ComputeStrategy, StrategySelector } from "./StrategySelector.js";

type CompleteStrategy = DeterminantStrategy & InverseStrategy;
export class DefaultBackend implements LinearAlgebraBackend {
    public constructor(private readonly strategy: CompleteStrategy) {
        if (strategy === null) throw new RangeError("strategy must implement DeterminantStrategy and InverseStrategy");
    }
    public id(): string {
        return this.strategy.id();
    }
    public determinantStrategy(): DeterminantStrategy {
        return this.strategy;
    }
    public inverseStrategy(): InverseStrategy {
        return this.strategy;
    }
}
export class LinearAlgebraEngine {
    private static defaultEngineInstance: LinearAlgebraEngine | null = null;
    public constructor(private readonly selectedBackend: LinearAlgebraBackend) {
        if (selectedBackend === null) throw new RangeError("backend cannot be null");
    }
    public static defaultEngine(): LinearAlgebraEngine {
        this.defaultEngineInstance ??= this.fromStrategy(ComputeStrategy.NAIVE_COFACTOR_CPU);
        return this.defaultEngineInstance;
    }
    public static fromStrategy(strategy: ComputeStrategy): LinearAlgebraEngine {
        return new LinearAlgebraEngine(new StrategySelector().select(strategy));
    }
    public determinant(matrix: MatrixNxM): number {
        return this.selectedBackend.determinantStrategy().determinant(matrix);
    }
    public inverse(matrix: MatrixNxM): MatrixNxM {
        return this.selectedBackend.inverseStrategy().inverse(matrix);
    }
    public backend(): LinearAlgebraBackend {
        return this.selectedBackend;
    }
}
